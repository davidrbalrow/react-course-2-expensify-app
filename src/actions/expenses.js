import uuid from 'uuid';
import database from '../firebase/firebase';
import { registerDecorator } from 'handlebars';
import { doesNotReject } from 'assert';

// conponent call action generator
// action generator returns object
// component dispactches object
// redux store chagnes

// component calls action generator
// aciton generator return function
// component dispatches function (?)
// function runs (dispatch other actions, and do whatever it wants)


// ADD_EXPENSE

export const addExpense = (expense) => ({
type: 'ADD_EXPENSE',
expense 
});

// START_ADD_EXPENSE


export const startAddExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            description = '', 
            note = '', 
            amount = 0, 
            createdAt = 0 
        } = expenseData;
      
        const expense = {description, note, amount, createdAt};

      return  database.ref(`/users/${uid}/expenses`).push(expense).then((ref)=>{
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};
// REMOVE_EXPENSE

export const removeExpense = ({id} = {}) => ({
type: 'REMOVE_EXPENSE',
id
});

// START_REMOVE_EXPENSE

export const startRemoveExpense = ({id} = {}) => {
    return (dispatch,getState) => {
    const uid = getState().auth.uid;
        return database.ref(`/users/${uid}/expenses/${id}`)
        .remove().then(()=>{       
        dispatch(removeExpense({id}));
                });    
    };
};

// EDIT_EXPENSE

export const editExpense = (id, updates) =>({
type: 'EDIT_EXPENSE',
id,
updates
});

//

export const startEditExpense = (id, updates) => {
    return (dispatch,getState) => {
        const uid = getState().auth.uid;
      return  database.ref(`/users/${uid}/expenses/${id}`)
      .update(updates).then(()=>{   
        dispatch(editExpense(id, updates));
        });
    };
};

// SET_EXPENSES
export const setExpenses = (expenses) => ({
    type:'SET_EXPENSES',
    expenses
});

//export const startSetExpenses;
export const startSetExpenses = () => {
    return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database.ref(`/users/${uid}/expenses`)
    .once('value',(snapshot)=>{
    const expenses = [];
    
    snapshot.forEach((childSnapshot)=>{
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val()  
        });
    });
  
    dispatch(setExpenses(expenses));
    
    });

                            
           
    };
};


//database.ref('expenses')
// database.ref('expenses')
// .on('value',(snapshot)=>{
//     const expenses = [];

//     snapshot.forEach((childSnapshot)=>{
//         expenses.push({
//           id: childSnapshot.apiKey,
//           ...childSnapshot.val()  
//         });
//     });
//     console.log(expenses);
// });

// dispatch(startSetExpenses({
//     expenses
// }));
        