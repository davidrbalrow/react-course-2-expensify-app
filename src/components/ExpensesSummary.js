import React from 'react';
import {connect} from 'react-redux';
import selectExpensesTotal from '../selectors/expenses-total';
import selectExpenses from '../selectors/expenses';

export const ExpensesSummary = (props) => (
   
    <div>
    <h3>Viewing {props.expenses.length} expense{props.expenses.length!==1 ? 's' : ''} totalling {selectExpensesTotal(props.expenses)}</h3>
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
}
 

export default connect(mapStateToProps)(ExpensesSummary);