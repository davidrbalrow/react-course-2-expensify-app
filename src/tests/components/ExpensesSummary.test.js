import React from 'react';
import { shallow } from 'enzyme';
import {ExpensesSummary} from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';
import toJSON from 'enzyme-to-json';

test('should render ExpensesSummary with expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenses={expenses}/>);
    expect(toJSON(wrapper)).toMatchSnapshot();
});

test('should render ExpensesSummary with no expenses', () => {
    const wrapper = shallow(<ExpensesSummary expenses={[]}/>);
    expect(toJSON(wrapper)).toMatchSnapshot();
});