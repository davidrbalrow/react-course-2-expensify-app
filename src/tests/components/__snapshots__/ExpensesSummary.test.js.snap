// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary with expenses 1`] = `
<div>
  <h3>
    Viewing 
    3
     expense
    s
     totalling 
    1950395
  </h3>
</div>
`;

exports[`should render ExpensesSummary with no expenses 1`] = `
<div>
  <h3>
    Viewing 
    0
     expense
    s
     totalling 
    0
  </h3>
</div>
`;
